src/negotiate.js

```diff
@@ -9,33 +9,82 @@ const mime = require("mime");
  * Express 4.x includes the the `res.format()` method, but it doesn't support
  * async handlers.
  *
+ * More than one handler for a type may be provided, in which case a stack of
+ * handlers is formed, like Express route handlers (but without any
+ * array-flattening behaviour).  In this case, all but the last handler in the
+ * stack should call "return next()" or "await next()" in order to chain to the
+ * next handler in the stack (unless that is not desired).
+ *
  * @param {Array[]}  providers      - Array of [Content-Type, handler] tuples
  * @param {String}   providers[][0] - Content-Type or shorthand like `json` or `html`
  * @param {Function} providers[][1] - Handler function that produces the associated Content-Type; may be async
  *
  * @returns {Function} An async routing function which will perform the dispatch.
  */
 function contentTypesProvided(providers) {
-  const types = providers.map(([type, handler]) => type); // eslint-disable-line no-unused-vars
-  const handlers = Object.fromEntries(providers);
+  const types = providers.map(([type, ...handlers]) => type); // eslint-disable-line no-unused-vars
+  const handlersByType = Object.fromEntries(providers.map(([type, ...handlers]) => [type, handlers]));
 
   return async function dispatchByContentType(req, res, next) {
+    // Check request Accept: against our Content-Type providers.
     res.vary("Accept");
 
+    // Automatically provide a Link: header for alternate types we support,
+    // even if we end up sending a 406 Not Acceptable.
```

Comment:
The concept of links seems really powerful, but with this implementation we are sending every possible alternate, regardless of whether that resource exists (e.g. `curl http://localhost:5000/zika  -I` indicates tip-frequencies as an alternate).

Is this a placeholder implementation, and one day we'll filter them using the `exists` method (e.g.), or is the idea that links are the set of all possible alternates and you need to try to get them to determine if they actually exist?

Reply:
Yeah, I think ideally we filter them in the future if feasible. For now (and maybe ~forever), it seems acceptable to me to expect clients to gracefully handle the 404 if they try to fetch a non-existent alternate. I didn't want to add filtering now since it'd be best with a stronger caching strategy when determining existence of sidecars, and it would add to the existing request amplification.

Reply:
👍